### PR TITLE
[FIXED JENKINS-47824] Looks like we have to do much more hoop jumping to correctly resolve things

### PR DESCRIPTION
**Pull request by stephenc**

### CODE REVIEW

**Comment by MarkEWaite** [comment]
@stephenc when I run this change in [my test case](https://github.com/MarkEWaite/jenkins-bugs/tree/JENKINS-47824) on my [docker instance](https://github.com/MarkEWaite/docker-lfs/tree/lts-with-plugins) it continues to fail with the message:

```
 > git fetch --no-tags --progress https://github.com/MarkEWaite/jenkins-pipeline-utils +refs/heads/${library.globalPipelineLibraryMarkEWaite.version}:refs/remotes/origin/${library.globalPipelineLibraryMarkEWaite.version} --depth=1
ERROR: Error cloning remote repo 'origin'
hudson.plugins.git.GitException: Command "git fetch --no-tags --progress https://github.com/MarkEWaite/jenkins-pipeline-utils +refs/heads/${library.globalPipelineLibraryMarkEWaite.version}:refs/remotes/origin/${library.globalPipelineLibraryMarkEWaite.version} --depth=1" returned status code 128:
stdout: 
stderr: fatal: Couldn't find remote ref refs/heads/v1.1

	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:1970)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:1689)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.access$300(CliGitAPIImpl.java:71)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$1.execute(CliGitAPIImpl.java:380)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$2.execute(CliGitAPIImpl.java:589)
	at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:1083)
	at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1123)
	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:113)
	at org.jenkinsci.plugins.workflow.libs.SCMSourceRetriever.doRetrieve(SCMSourceRetriever.java:112)
	at org.jenkinsci.plugins.workflow.libs.SCMRetriever.retrieve(SCMRetriever.java:64)
	at org.jenkinsci.plugins.workflow.libs.LibraryAdder.retrieve(LibraryAdder.java:153)
	at org.jenkinsci.plugins.workflow.libs.LibraryAdder.add(LibraryAdder.java:134)
	at org.jenkinsci.plugins.workflow.libs.LibraryDecorator$1.call(LibraryDecorator.java:125)
	at org.codehaus.groovy.control.CompilationUnit.applyToPrimaryClassNodes(CompilationUnit.java:1065)
	at org.codehaus.groovy.control.CompilationUnit.doPhaseOperation(CompilationUnit.java:603)
	at org.codehaus.groovy.control.CompilationUnit.processPhaseOperations(CompilationUnit.java:581)
	at org.codehaus.groovy.control.CompilationUnit.compile(CompilationUnit.java:558)
	at groovy.lang.GroovyClassLoader.doParseClass(GroovyClassLoader.java:298)
	at groovy.lang.GroovyClassLoader.parseClass(GroovyClassLoader.java:268)
	at groovy.lang.GroovyShell.parseClass(GroovyShell.java:688)
	at groovy.lang.GroovyShell.parse(GroovyShell.java:700)
	at org.jenkinsci.plugins.workflow.cps.CpsGroovyShell.doParse(CpsGroovyShell.java:129)
	at org.jenkinsci.plugins.workflow.cps.CpsGroovyShell.reparse(CpsGroovyShell.java:123)
	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.parseScript(CpsFlowExecution.java:517)
	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.start(CpsFlowExecution.java:480)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun.run(WorkflowRun.java:269)
	at hudson.model.ResourceController.execute(ResourceController.java:97)
	at hudson.model.Executor.run(Executor.java:421)
ERROR: Error cloning remote repo 'origin'
org.codehaus.groovy.control.MultipleCompilationErrorsException: startup failed:
WorkflowScript: Loading libraries failed

1 error

	at org.codehaus.groovy.control.ErrorCollector.failIfErrors(ErrorCollector.java:310)
	at org.codehaus.groovy.control.CompilationUnit.applyToPrimaryClassNodes(CompilationUnit.java:1085)
	at org.codehaus.groovy.control.CompilationUnit.doPhaseOperation(CompilationUnit.java:603)
	at org.codehaus.groovy.control.CompilationUnit.processPhaseOperations(CompilationUnit.java:581)
	at org.codehaus.groovy.control.CompilationUnit.compile(CompilationUnit.java:558)
	at groovy.lang.GroovyClassLoader.doParseClass(GroovyClassLoader.java:298)
	at groovy.lang.GroovyClassLoader.parseClass(GroovyClassLoader.java:268)
	at groovy.lang.GroovyShell.parseClass(GroovyShell.java:688)
	at groovy.lang.GroovyShell.parse(GroovyShell.java:700)
	at org.jenkinsci.plugins.workflow.cps.CpsGroovyShell.doParse(CpsGroovyShell.java:129)
	at org.jenkinsci.plugins.workflow.cps.CpsGroovyShell.reparse(CpsGroovyShell.java:123)
	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.parseScript(CpsFlowExecution.java:517)
	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.start(CpsFlowExecution.java:480)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun.run(WorkflowRun.java:269)
	at hudson.model.ResourceController.execute(ResourceController.java:97)
	at hudson.model.Executor.run(Executor.java:421)
Finished: FAILURE
```

**Comment by MarkEWaite** [comment]
Steps to duplicate my setup where I see the problem:

```
#!/bin/bash
git lfs clone https://github.com/MarkEWaite/docker-lfs JENKINS-47824
cd JENKINS-47824
git lfs fetch origin origin/lts-with-plugins
git checkout -b lts-with-plugins -t origin/lts-with-plugins
docker build -t jenkins-47824:bad .
docker run --rm -i -p 9090:8080 -t jenkins-47824:bad
```

With that docker image running on [port 9090](http://mark-pc2:9090/), open the [Bugs-Pipeline Checks / jenkins-bugs folder](http://mark-pc2:9090/job/Bugs-Pipeline-Checks/job/jenkins-bugs-multibranch-pipeline/) and click "[scan multibranch pipeline now](http://localhost:9090/job/Bugs-Pipeline-Checks/job/jenkins-bugs-multibranch-pipeline/build?delay=0)"

**Comment by stephenc** [comment]
Will also nee https://github.com/jenkinsci/github-branch-source-plugin/pull/166 if the share library source is GitHub

File: src/main/java/jenkins/plugins/git/AbstractGitSCMSource.java [hunk 1]
```diff
@@ -688,23 +689,146 @@ protected SCMRevision retrieve(@NonNull final String revision, @NonNull final Ta
...
-                                      } catch (GitException x2) {
-                                          listener.getLogger().println(x.getMessage());
-                                          listener.getLogger().println(x2.getMessage());
+                                      objectId = client.revParse(revision);
+                                      hash = objectId.name();
+                                      List<Branch> branches = client.getBranchesContaining(hash, false);
+                                      if (branches.isEmpty()) {
+                                          listener.getLogger().printf("Could not find a branch containing commit %s%n",
+                                                  hash);
                                           return null;
```
Comment (jglick) [inline src/main/java/jenkins/plugins/git/AbstractGitSCMSource.java]: #557

**Pull request merged**